cpp/src/quantiles/quantiles.cu

```diff
@@ -61,48 +60,216 @@ quantiles(table_view const& input,
+                                              stream,
+                                              mr);
+
+        if (output->size() == 0) {
+            return output;
+        }
+
+        auto d_input = column_device_view::create(input);
+        auto d_output = mutable_column_device_view::create(output->mutable_view());
+
+        rmm::device_vector<double> q_device{q};
+
+        auto sorted_data = thrust::make_permutation_iterator(input.data<T>(),
+                                                             ordered_indices);
 
+        if (retain_types)
+        {
+            thrust::transform(
+                q_device.begin(),
+                q_device.end(),
+                d_output->template begin<T>(),
+                [sorted_data, interp=interp, size=input.size()]
+                __device__ (double q){
+                    return select_quantile_data<T>(sorted_data, size, q, interp);
+                });
+        }
+        else
+        {
+            thrust::transform(
+                q_device.begin(),
+                q_device.end(),
+                d_output->template begin<double>(),
+                [sorted_data, interp=interp, size=input.size()]
+                __device__ (double q){
+                    return select_quantile_data<double>(sorted_data, size, q, interp);
+                });
+        }
+
+        if (input.nullable())
+        {
+            auto sorted_validity = thrust::make_transform_iterator(
+                ordered_indices,
+                [input=*d_input] __device__ (size_type idx) {
+                    return input.is_valid_nocheck(idx);
+                });
+
+            rmm::device_buffer mask;
+            size_type null_count;
+
+            std::tie(mask, null_count) = valid_if(
+                q_device.begin(),
+                q_device.end(),
+                [sorted_validity, size=input.size(), interp=interp]
+                __device__(double q) {
+                    return select_quantile_validity(sorted_validity,
+                                                    size,
+                                                    q,
+                                                    interp);
+                },
+                stream,
+                mr);
+
+            output->set_null_mask(std::move(mask), null_count);
+        }
+
+        return output;
+    }
+};
+
+template<typename SortMapIterator>
 std::unique_ptr<table>
 quantiles(table_view const& input,
+          SortMapIterator ordered_indices,
           std::vector<double> const& q,
           interpolation interp,
-          cudf::sorted is_input_sorted,
-          std::vector<order> const& column_order,
-          std::vector<null_order> const& null_precedence,
+          bool retain_types,
           rmm::mr::device_memory_resource* mr)
 {
-    CUDF_FUNC_RANGE();
-    if (q.size() == 0) {
-        return empty_like(input);
+    auto is_input_numeric = all_of(input.begin(),
```

Comment:
```suggestion
    auto is_input_numeric = std::all_of(input.begin(),
```

Reply:
Good catch. Thanks!

Reply:
I'm confused as to how this even compiled. Is there a `using namespace std` somewhere?

Reply:
Not afaik. i haven't investigated, but maybe `cudf`, `experimental`, or `detail` has a compatible `all_of` ?